src/topology/perfect.lean

```diff
@@ -0,0 +1,279 @@
+/-
+Copyright (c) 2022 Felix Weilacher. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Felix Weilacher
+-/
+import topology.separation
+import topology.bases
+
+/-!
+# Perfect Sets
+
+In this file we define perfect subsets of a topological space, and prove some basic properties,
+including a version of the Cantor-Bendixson Theorem.
+
+## Main Definitions
+
+* `acc_pt x C`: A point `x` is an accumulation point of a set `C`,
+  taken here to mean that every neighborhood of `x` contains points of `C` other than `x`
+* `perfect C`: A set `C` is perfect, meaning it is closed and every point of it
+  is an accumulation point of itself.
+
+## Main Statements
+
+* `perf_nonempty.splitting`: A perfect nonempty set contains two disjoint perfect nonempty subsets.
+  The main inductive step in the construction of an embedding from the Cantor space to a
+  perfect nonempty complete metric space.
+* `ctble_union_perfect_of_closed`: One version of the Cantor-Bendixson Theorem: A closed
+  set in a second countable space can be written as the union of a countable set and a perfect set.
+
+## Implementation Notes
+
+We do not require perfect sets to be nonempty, a condition which is often part of
+the definition of perfect. We include an extra definition `perf_nonempty`, which bundles
+these two conditions.
+
+We define a nonstandard predicate, `preperfect`, which drops the closed-ness requirement
+from the definition of perfect. In T1 spaces, this is equivalent to having a perfect closure,
+see `preperfect_iff_perfect_closure`.
+
+## References
+
+* [kechris1995] (Chapter 6)
+
+## Tags
+
+accumulation point, perfect set, Cantor-Bendixson.
+
+-/
+
+section
+
+open_locale topological_space filter
+open topological_space filter set
+
+variables {α : Type*} [topological_space α]
+
+/-- A point `x` is an accumulation point of a set `C` if every neighborhood
+of `x` contains a point of `C` other than `x`. See `acc_pt_iff`. -/
+def acc_pt (x : α) (C : set α) : Prop := ne_bot (𝓝[≠] x ⊓ 𝓟 C)
```

Comment:
I would suggest generalizing this to filter (like `cluster_pt`), and then use it with principal filters to talk about accumulation points of sets. This should also probably be moved near `cluster_pt`.

Reply:
done